app/services/checkout_service.rb

```diff
@@ -0,0 +1,51 @@
+class CheckoutService
+  def initialize(current_user, order, stripe_token)
+    @current_user = current_user
+    @order = order
+    @stripe_token = stripe_token
+  end
+
+  def place_order
+    customer_id = retrieve_stripe_customer_id
+    @current_user.update_attribute(:stripe_customer_id, customer_id)
+    charge_customer(customer_id)
+    finalize
+    send_invoice
+  end
+
+  private
+
+  def retrieve_stripe_customer_id
+    if @stripe_token
+      customer = Stripe::Customer.create(
+        email:  @current_user.email,
+        source: @stripe_token
+      )
+      return customer.id
+    else
+      return @current_user.stripe_customer_id
+    end
+  end
+
+  def charge_customer(customer_id)
+    Stripe::Charge.create(
+      customer:     customer_id,
+      amount:       @order.stripe_total,
+      description:  'The Beautiful Rails Bookstore Purchase',
+      currency:     'usd'
+    )
+
+  rescue Stripe::CardError => e
+    flash[:error] = e.message
+    redirect_to new_charge_path
```

Comment:
When breaking functionality out, it is important to think of what responsibilities each piece of your software has. In general, a module like this in the service layer should act on data, but not modify the http response since that is the job of the controller. Instead of having handling the redirect logic here, consider having this service return whether it was successful or not and having the controller handle where to go so that the response logic is all in one place.